xarray/test/test_dataarray.py

```diff
@@ -1791,3 +1791,41 @@ def test_full_like(self):
         actual = _full_like(DataArray([1, 2, 3]), fill_value=np.nan)
         self.assertEqual(actual.dtype, np.float)
         np.testing.assert_equal(actual.values, np.nan)
+    
+    def test_dot(self):
+        x = np.linspace(-3, 3, 6)
+        y = np.linspace(-3, 3, 5)
+        z = range(4) 
+        da_vals = np.arange(6 * 5 * 4).reshape((6, 5, 4))
+        da = DataArray(da_vals, coords=[x, y, z], dims=['x', 'y', 'z'])
+        
+        dm_vals = range(4)
+        dm = DataArray(dm_vals, coords=[z], dims=['z'])
+        
+        #nd dot 1d
+        actual = da.dot(dm)
+        expected_vals = np.tensordot(da_vals, dm_vals, [2, 0])
+        expected = DataArray(expected_vals, coords=[x, y], dims=['x', 'y'])
+        self.assertDataArrayEqual(expected, actual)
+        
+        #all shared dims
+        actual = da.dot(da)
+        expected_vals = np.tensordot(da_vals, da_vals, axes=([0, 1, 2], [0, 1, 2]))
+        expected = DataArray(expected_vals)
+        self.assertDataArrayEqual(expected, actual)
+        
+        #multiple shared dims
+        dm_vals = np.arange(20 * 5 * 4).reshape((20, 5, 4))
+        j = np.linspace(-3, 3, 20)
+        dm = DataArray(dm_vals, coords=[j, y, z], dims=['j', 'y', 'z'])
+        actual = da.dot(dm)
+        expected_vals = np.tensordot(da_vals, dm_vals, axes=([1, 2], [1, 2]))
+        expected = DataArray(expected_vals, coords=[x, j], dims=['x', 'j'])
+        self.assertDataArrayEqual(expected, actual)
+        
+        with self.assertRaises(NotImplementedError):
+            da.dot(dm.to_dataset(name='dm'))
+        with self.assertRaises(TypeError):
+            da.dot(dm.values)
+        with self.assertRaises(ValueError):
```

Comment:
use something like `self.assertRaisesRegexp(ValueError, 'no shared dimensions')` to verify that it's the right `ValueError` (there are a lot of ways to get a ValueError!)